docs/source/nested.rst

```diff
@@ -10,213 +10,471 @@ Introduction
+         [ 4.2000,  4.2000,  4.2000],
+         [ 4.2000,  4.2000,  4.2000],
+         [ 4.2000,  4.2000,  4.2000]],
+        [[-0.0479, -0.7610, -0.3484],
+         [ 1.1345,  1.0556,  0.3634],
+         [-1.7122, -0.5921,  0.0540],
+         [-0.5506,  0.7608,  2.0606],
+         [ 1.5658, -1.1934,  0.3041],
+         [ 0.1483, -1.1284,  0.6957]]])
+
+This can be useful as an escape hatch to work around NJT support gaps, but ideally such
+conversions should be avoided when possible for optimal memory usage and performance, as the
+more efficient nested tensor layout does not materialize padding.
+
+The reverse conversion can be accomplished using ``torch.nested.narrow()``, which applies
+ragged structure to a given dense tensor to produce an NJT. Note that by default, this operation
+does not copy the underlying data, and thus the output NJT is generally non-contiguous. It may be
+useful to explicitly call ``contiguous()`` here if a contiguous NJT is desired.
+
+>>> padded = torch.randn(3, 5, 4)
+>>> seq_lens = torch.tensor([3, 2, 5], dtype=torch.int64)
+>>> nt = torch.nested.narrow(padded, dim=1, length=seq_lens, layout=torch.jagged)
+>>> nt.shape
+torch.Size([3, j1, 4])
+>>> nt = nt.contiguous()
+>>> nt.shape
+torch.Size([3, j2, 4])
+
+Shape manipulations
+^^^^^^^^^^^^^^^^^^^
+
+Nested tensors support a wide array of operations for shape manipulation, including views.
+
+>>> a = torch.randn(2, 6)
+>>> b = torch.randn(4, 6)
+>>> nt = torch.nested.nested_tensor([a, b], layout=torch.jagged)
+>>> nt.shape
+torch.Size([2, j1, 6])
+>>> nt.unsqueeze(-1).shape
+torch.Size([2, j1, 6, 1])
+>>> nt.unflatten(-1, [2, 3]).shape
+torch.Size([2, j1, 2, 3])
+>>> torch.cat([nt, nt], dim=2).shape
+torch.Size([2, j1, 12])
+>>> torch.stack([nt, nt], dim=2).shape
+torch.Size([2, j1, 2, 6])
+>>> nt.transpose(-1, -2).shape
+torch.Size([2, 6, j1])
+
+Attention mechanisms
+^^^^^^^^^^^^^^^^^^^^
+
+As variable-length sequences are common inputs to attention mechanisms, nested tensors support
+important attention operators
+`Scaled Dot Product Attention (SDPA) <https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html>`_ and
+`FlexAttention <https://pytorch.org/docs/stable/nn.attention.flex_attention.html#module-torch.nn.attention.flex_attention>`_.
+See
+`here <https://pytorch.org/tutorials/intermediate/transformer_building_blocks.html#multiheadattention>`__
+for usage examples of NJT with SDPA and
+`here <https://pytorch.org/tutorials/intermediate/transformer_building_blocks.html#flexattention-njt>`__
+for usage examples of NJT with FlexAttention.
+
+.. _usage_with_torch_compile:
+
+Usage with torch.compile
+++++++++++++++++++++++++
+
+NJTs are designed to be used with ``torch.compile()`` for optimal performance. In particular,
+``torch.compile()`` is important to eliminate eager-mode python tensor subclass overhead, which
+affects NJT as it is a python tensor subclass. NJTs work out-of-the-box and
```

Comment:
I wonder if it is worth mentioning this explicitly, since maybe that is the expectation already when working with compile.

Reply:
true, it should be the expectation. I'm making it explicit to be clear about what users should expect from NJT. There are other features within PyTorch that they can't expect to compose with torch.compile without issue, so it's nice to set this one apart via "you should expect a seamless experience with torch.compile"